### PR TITLE
[BOLT][DWARF] Add support for transitive DW_AT_name/DW_AT_linkage_name resolution for DW_AT_name/DW_AT_linkage_name.

### PR DESCRIPTION
**Pull request by ayermolo**
This fix handles a case where a DIE that does not have DW_AT_name/DW_AT_linkage_name, but has a reference to another DIE using  DW_AT_abstract_origin/DW_AT_specification. It also fixes a bug where there are cross CU references for those attributes. Previously it would use a DWARF Unit of a DIE which was being processed The warf5-debug-names-cross-cu.s test just happened to work because how it was constructed where string section was shared by both DWARF Units.

To resolve DW_AT_name/DW_AT_linkage_name this patch iterates over references until it either reaches the final DIE or finds both of those names.

### CODE REVIEW

**Review by dcci — COMMENTED** [review]
Looks fine to me, but please change the commit message to match the title rather than using `refactor + support abstr chain`

**Review by dcci — APPROVED** [review]

File: bolt/lib/Core/DebugNames.cpp [hunk 1]
```diff
@@ -222,134 +222,114 @@ static uint64_t getEntryID(const BOLTDWARF5AccelTableData &Entry) {
...
-      It.HashValue = caseFoldingDjbHash(Name);
+      else
+        NameIndexOffset = ItCache->second;
     }
-    return Name;
-  };
+    if (!NameToUse.empty())
+      NameIndexOffset = MainBinaryStrWriter.addString(Name);
+    It.StrOffset = NameIndexOffset;
+    // This the same hash function used in DWARF5AccelTableData.
```
Comment (dcci) [inline bolt/lib/Core/DebugNames.cpp:273]: "this is"

File: bolt/lib/Core/DebugNames.cpp [hunk 2]
```diff
@@ -359,24 +339,97 @@ DWARF5AcceleratorTable::addAccelTableEntry(
...
+    const uint32_t NumberParentsInChain,
+    std::optional<BOLTDWARF5AccelTableData *> &Parent) {
+  if (Unit.getVersion() < 5 || !NeedToCreate)
+    return std::nullopt;
+  std::string NameToUse = "";
+
+  if (!canProcess(Unit, Die, NameToUse, false))
+    return std::nullopt;
+
+  // Addes a Unit to either CU, LocalTU or ForeignTU list the first time we
```
Comment (dcci) [inline bolt/lib/Core/DebugNames.cpp:390]: Addes -> adds.

File: bolt/lib/Core/DebugNames.cpp [hunk 3]
```diff
@@ -359,24 +339,97 @@ DWARF5AcceleratorTable::addAccelTableEntry(
...
+
+  std::optional<BOLTDWARF5AccelTableData *> NameEntry =
+      addEntry(Unit, Die, DWOID, Parent,
+               getName(Unit, DWOID, NameToUse, NameVal), NumberParentsInChain);
+  if (NameEntry)
+    return NameEntry;
+
+  // The DIE doesn't have DW_AT_name or DW_AT_linkage_name, so we need to see if
+  // we can follow other attributes to find them. For the purposes of
+  // determining whether a debugging information entry has a particular
```
Comment (dcci) [inline bolt/lib/Core/DebugNames.cpp:417]: debugging -> debug.

**Comment by ayermolo** [comment]
> Looks fine to me, but please change the commit message to match the title rather than using `refactor + support abstr chain`

That will be automatically be used when Squash and merge through UI.
Thanks for review.

**Review by dcci — COMMENTED** [review]
LGTM

**Pull request merged**